[WIP] Add HSDP dtensor state_dict support

wz337 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105990

This is the diff train to enable HSDP DTensor state_dict support. In this PR,

1. We add _device_mesh to FSDP kwarg
2. Validate process_group and _device_mesh is mutually exclusive.
3. When _device_mesh is passed in, we re-use the PG from DeviceMesh for both 1D FSDP and 2D HSDP.
4. Utilize the _device_mesh to do operations for state_dict and optim_state_dict manipulation.

wz337 committed:
1_n_add_device_mesh_and_validate

[ghstack-poisoned]

wz337 committed:
Update on "[RFC][1/N] add _device_mesh to FSDP kwarg"


This is the diff train to enable HSDP DTensor state_dict support. In this PR,

1. We add _device_mesh to FSDP kwarg
2. Validate process_group and _device_mesh is mutually exclusive.
3. When _device_mesh is passed in, we re-use the PG from DeviceMesh for both 1D FSDP and 2D HSDP.

[ghstack-poisoned]

rohan-varma commented on test/distributed/fsdp/test_fsdp_dtensor_state_dict.py:
```diff
@@ -227,6 +228,30 @@ def test_dtensor_sharded_model_load_state_dict(self, offload_to_cpu):
+            sharding_strategy=ShardingStrategy.HYBRID_SHARD,
+        )
+        optim = torch.optim.Adam(model.parameters(), lr=0.1)
+        model(model.get_input()).sum().backward()
```
nit (i know the PR is WIP) ideally the test should validate something like we're sharded / not sharded over the right dimensions?

rohan-varma commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -98,21 +99,34 @@ def _init_process_group_state(
+    device_mesh: DeviceMesh = None,
 ) -> _FSDPState:
+    if process_group is not None and device_mesh is not None:
+        raise ValueError(
```
nit: add a test for this

rohan-varma reviewed (commented):
per discussion with @fegin, it might be important for PT-D to ensure composable API stays in sync with FSDP as a first-class citizen. Might be good to sync w/ @fegin to see if we should include composable API changes in this PR as well.

wz337 commented on test/distributed/fsdp/test_fsdp_dtensor_state_dict.py:
```diff
@@ -227,6 +228,30 @@ def test_dtensor_sharded_model_load_state_dict(self, offload_to_cpu):
+        device_mesh = DeviceMesh(self.device_type, mesh_tensor)
+        model = FSDP(
+            TestDummyModel().cuda(),
+            _device_mesh=device_mesh,
```
I am hardcoding the mesh dim for HSDP sharded and replicate for now, since we want to quickly roll this part out so modelstore folks can start build the integration for sharded DTensor as well 2D device-mesh support. 

For users, they actually don't have to know the mesh dimension for dp replicate and dp shard, since the default would be dp_replicate: 0, dp_shard: 1. For more advanced users, they can pass an additional arguments in initializing device_mesh. More concrete example is in this diff Chien-Chin has. D47781455

wz337 committed:
Update on "[RFC][1/N] add _device_mesh to FSDP kwarg"


This is the diff train to enable HSDP DTensor state_dict support. In this PR,

1. We add _device_mesh to FSDP kwarg
2. Validate process_group and _device_mesh is mutually exclusive.
3. When _device_mesh is passed in, we re-use the PG from DeviceMesh for both 1D FSDP and 2D HSDP.

[ghstack-poisoned]

wz337 committed:
Update on "[WIP] Add HSDP dtensor state_dict support"


This is the diff train to enable HSDP DTensor state_dict support. In this PR,

1. We add _device_mesh to FSDP kwarg
2. Validate process_group and _device_mesh is mutually exclusive.
3. When _device_mesh is passed in, we re-use the PG from DeviceMesh for both 1D FSDP and 2D HSDP.
4. Utilize the _device_mesh to do operations for state_dict and optim_state_dict manipulation.

[ghstack-poisoned]

wz337 committed:
Update on "[WIP] Add HSDP dtensor state_dict support"


This is the diff train to enable HSDP DTensor state_dict support. In this PR,

1. We add _device_mesh to FSDP kwarg
2. Validate process_group and _device_mesh is mutually exclusive.
3. When _device_mesh is passed in, we re-use the PG from DeviceMesh for both 1D FSDP and 2D HSDP.
4. Utilize the _device_mesh to do operations for state_dict and optim_state_dict manipulation.

[ghstack-poisoned]

wz337 committed:
Update on "[WIP] Add HSDP dtensor state_dict support"


This is the diff train to enable HSDP DTensor state_dict support. In this PR,

1. We add _device_mesh to FSDP kwarg
2. Validate process_group and _device_mesh is mutually exclusive.
3. When _device_mesh is passed in, we re-use the PG from DeviceMesh for both 1D FSDP and 2D HSDP.
4. Utilize the _device_mesh to do operations for state_dict and optim_state_dict manipulation.

[ghstack-poisoned]

wz337 committed:
Update on "[WIP] Add HSDP dtensor state_dict support"


This is the diff train to enable HSDP DTensor state_dict support. In this PR,

1. We add _device_mesh to FSDP kwarg
2. Validate process_group and _device_mesh is mutually exclusive.
3. When _device_mesh is passed in, we re-use the PG from DeviceMesh for both 1D FSDP and 2D HSDP.
4. Utilize the _device_mesh to do operations for state_dict and optim_state_dict manipulation.

[ghstack-poisoned]

wz337 committed:
Update on "[WIP] Add HSDP dtensor state_dict support"


This is the diff train to enable HSDP DTensor state_dict support. In this PR,

1. We add _device_mesh to FSDP kwarg
2. Validate process_group and _device_mesh is mutually exclusive.
3. When _device_mesh is passed in, we re-use the PG from DeviceMesh for both 1D FSDP and 2D HSDP.
4. Utilize the _device_mesh to do operations for state_dict and optim_state_dict manipulation.

[ghstack-poisoned]

wz337 committed:
Update on "[WIP] Add HSDP dtensor state_dict support"


This is the diff train to enable HSDP DTensor state_dict support. In this PR,

1. We add _device_mesh to FSDP kwarg
2. Validate process_group and _device_mesh is mutually exclusive.
3. When _device_mesh is passed in, we re-use the PG from DeviceMesh for both 1D FSDP and 2D HSDP.
4. Utilize the _device_mesh to do operations for state_dict and optim_state_dict manipulation.

[ghstack-poisoned]

wz337 commented:
close as we have https://github.com/pytorch/pytorch/pull/107533
